KEYCLOAK-11511 Improve exception handling of REST user creation

k-tamura opened the pull request:
Please refer to [KEYCLOAK-11511](https://issues.jboss.org/browse/KEYCLOAK-11511).

Cracky5457 commented on services/src/main/java/org/keycloak/credential/PasswordCredentialProvider.java:
```diff
@@ -164,7 +164,7 @@ public void disableCredentialType(RealmModel realm, UserModel user, String crede
     @Override
     public boolean supportsCredentialType(String credentialType) {
-        return credentialType.equals(CredentialModel.PASSWORD);
+        return credentialType == null || credentialType.equals(CredentialModel.PASSWORD);
```
FYI ` CredentialModel.PASSWORD.equals(credentialType)` in that order can't throw a null pointer exception, usually always the constants first because equals accept "null" but null don't have equals function.

k-tamura replied:
I think _true_ should be returned if _credentialType_ is null, that is _credentialType_ is not explicitly specified. In other words, default _credentialType_ is _CredentialModel.PASSWORD_.

Cracky5457 replied:
It what I mean, that was just fyi, `CredentialModel.PASSWORD.equals(credentialType)` is the very same thing than `credentialType == null || credentialType.equals(CredentialModel.PASSWORD);` No changes required tho, it's just shorter.

Cracky5457 replied:
Ok, I see what you mean, I though you was trying to avoid a null pointer exception.

Cracky5457 commented on services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:
```diff
@@ -104,17 +105,20 @@ public UsersResource(RealmModel realm, AdminPermissionEvaluator auth, AdminEvent
         auth.users().requireManage();
-
+        String username = rep.getUsername();
+        if (username == null || username.length() == 0) {
```
`username.trim().length() == 0` maybe if you don't want blank user

k-tamura replied:
Just as you said. I will fix so.

hmlnarik commented on services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:
```diff
@@ -131,12 +135,17 @@ public Response createUser(final UserRepresentation rep) {
             }
             logger.warn("Could not create user", me);
-            return ErrorResponse.exists("Could not create user");
+            return ErrorResponse.error("Could not create user", Response.Status.INTERNAL_SERVER_ERROR);
```
```suggestion
            return ErrorResponse.error("Could not create user", Response.Status.BAD_REQUEST);
```

hmlnarik reviewed (changes requested):
Please remove the change in PasswordCredentialProvider, the credentialType is mandatory. Once you would add tests to validate various failure scenarios (see `org.keycloak.testsuite.admin.UserTest`), I believe this can be merged.

k-tamura commented:
> Please remove the change in PasswordCredentialProvider, the credentialType is mandatory.

If the change is removed, user creation without credential type ( `"type" : "password",` ) causes NPE as follows:

```
curl --dump-header -  -X POST \
 -H "Authorization: Bearer eyJ..." \
 -H "Content-Type: application/json; charset=utf-8" \
 -d '{"username" : "test", "email" : "test1@gmail.com", "credentials" : [{"value" : "password"}]}' \
 "http://localhost:8080/auth/admin/realms/master/users"
HTTP/1.1 500 Internal Server Error
  ...
```


```
16:13:31,422 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (default task-63) Uncaught server error: java.lang.NullPointerException
        at org.keycloak.credential.PasswordCredentialProvider.supportsCredentialType(PasswordCredentialProvider.java:167)
        at org.keycloak.credential.UserCredentialStoreManager.updateCredential(UserCredentialStoreManager.java:184)
        at org.keycloak.models.utils.RepresentationToModel.updateCredential(RepresentationToModel.java:1605)
        at org.keycloak.models.utils.RepresentationToModel.createCredentials(RepresentationToModel.java:1590)
        at org.keycloak.services.resources.admin.UsersResource.createUser(UsersResource.java:121)
        at sun.reflect.GeneratedMethodAccessor774.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
```

I think NPE should be avoided at least, for example:

```java
if (credentialType == null) throw new ModelException("Credential type is required");
```

> Once you would add tests to validate various failure scenarios (see org.keycloak.testsuite.admin.UserTest), I believe this can be merged.

OK. I will write tests. Please wait a while.

hmlnarik commented:
> ```java
> if (credentialType == null) throw new ModelException("Credential type is required");
> ```

Yes. This check should be placed into the beginning of `RepresentationToModel.updateCredential` method.

k-tamura commented:
@hmlnarik Could you review my latest commit? I added four test cases and introduce `ObjectUtil.isBlank()` instead of null checking.

Note: I have not rebased (not merged my commits all in one) yet so that it is easier to review. If needed, I will rebase.

stianst commented:
This will also resolve https://issues.jboss.org/browse/KEYCLOAK-11974

maksymgendin commented on services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:
```diff
@@ -131,12 +136,17 @@ public Response createUser(final UserRepresentation rep) {
+            if (session.getTransactionManager().isActive()) {
+                session.getTransactionManager().setRollbackOnly();
+            }
+            return ErrorResponse.error("Password policy not met for user " + username, Response.Status.BAD_REQUEST);
```
@k-tamura I have some privacy/security concerns in returning the username in the response body. Maybe it is a better idea not to provide it there?

k-tamura replied:
@maksymgendin
Umm... I think there are no security/privacy problems but user name does not need to be included in response. So I will remove it if you have any concerns. 

@stianst @hmlnarik 
How do you think?

mposolda replied:
@k-tamura My vote is also to not include username in the response

mposolda commented on services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:
```diff
@@ -131,12 +136,17 @@ public Response createUser(final UserRepresentation rep) {
             }
             logger.warn("Could not create user", me);
-            return ErrorResponse.exists("Could not create user");
+            return ErrorResponse.error(me.getMessage(), Response.Status.BAD_REQUEST);
```
It will be good to display some generic error message, but not add exception message to the response. In case that exception message would contain some user-provided data and the error response is later shown on the HTML page, various kind of attacks (eg XSS) can happen

mposolda reviewed (changes requested):
@k-tamura The PR looks like quite good, but few updates might be still needed. It seems that:
- After added check in RepresentationToModel class, changes in the PasswordCredentialProvider should be removed now
- The error responses should contain generic error messages and not show username or any other data user-provided data. There are 2 places in the PR, which are commented in this manner.
- The rebase will be needed

mposolda commented:
Added label "Hold" as this is not a blocker for Keycloak 9.0.0

k-tamura commented:
OK. I will do next week.

k-tamura committed:
KEYCLOAK-11511 Improve exception handling of REST user creation

k-tamura commented:
Hi Keycloak dev team,

Please check my latest commit.

As a side note, I removed the code and its test code:

```java
if (credentialType == null) throw new ModelException("Credential type is required");
```
I added  the code to avoid NPE, but it is no longer needed since NPE has been fixed in [KEYCLOAK-11745](https://github.com/keycloak/keycloak/pull/6459).

Best regards,

mposolda reviewed (approved):
@k-tamura Thanks! PR looks good to me now and I am approving. 
@hmlnarik Feel free to review and merge whenever you think this is ready to go. I removed myself as an assignee for this.

hmlnarik reviewed (approved):
Looks good to me, thank you, @k-tamura !
